cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java

```diff
@@ -336,9 +338,112 @@ public void rollback(TopicId topicId, RollbackDetail rollbackDetail)
     throw new IOException("NOT IMPLEMENTED");
   }
 
+  /**
+   * Fetches messages from the specified topic.
+   *
+   * <p>The fetch operation utilizes the following protocol:
+   *     <ul>
+   *         <li>**MessageId Construction:**
+   *             MessageId is constructed as the concatenation of `sequence_id` and `publish_ts`.
+   *             This guarantees the following: If the `publish_ts` of message A is less than the
+   *            `publish_ts` of message B, then message A is delivered before message B.
+   *         </li>
+   *         <li>**Fetch Logic:**
+   *             - If the provided `messageId` exists in-memory, the iterator pointing to the next message is returned.
+   *             - Otherwise, a Spanner read operation is issued to fetch messages
+   *               with `publish_ts` greater than the `publish_ts` in the provided `messageId`.
+   *             - Upon reading rows, chunked messages are reconstructed as a single message and are held in memory.
+   *         </li>
+   *     </ul>
+   * </p>
+   */
   @Override
   public CloseableIterator<RawMessage> fetch(MessageFetchRequest messageFetchRequest)
       throws TopicNotFoundException, IOException {
-    throw new IOException("NOT IMPLEMENTED");
+    // If the client is fetching for the first time ever,
+    // they should fetch data for publish_ts > 0 and sequence_id > -1.
+    Long startTime = 0L;
+    if (messageFetchRequest.getStartTime() != null) {
+      startTime = messageFetchRequest.getStartTime();
+    }
+    short sequenceId = -1;
+    // MessageId is constructed as the concatenation of SequenceNo and CommitTs columns.
+    // These fields are set in the offset field of the message fetch request.
+    byte[] id = messageFetchRequest.getStartOffset();
+    if (id != null) {
+      int offset = 0;
+      startTime = Bytes.toLong(id, offset);
+      offset += Bytes.SIZEOF_LONG;
+      sequenceId = Bytes.toShort(id, offset);
+    }
+
+    LOG.trace("Fetching message from topic : {} with start time : {} sequenceId : {}",
+        messageFetchRequest.getTopicId().getTopic(), startTime, sequenceId);
+    // publish_ts > TIMESTAMP_MICROS(startTime)
+    // or
+    // publish_ts = TIMESTAMP_MICROS(startTime) and sequence_id > sequenceId
+    // order by
+    // publish_ts, sequence_id
+    String sqlStatement = String.format(
+        "SELECT %s, %s, UNIX_MICROS(%s), %s FROM %s where (%s > TIMESTAMP_MICROS(%s)) or"
+            + " (%s = TIMESTAMP_MICROS(%s) and %s > %s) order by" + " %s, %s LIMIT %s",
```

Comment:
Did we confirm this would use indexes? I am also a bit worried we are doing `%s = TIMESTAMP_MICROS(%s)`. We essentially do truncation and then doing an "equal" comparison of this truncation results. It may lead to subtle errors unless we are strict about sequence ids.
E.g. let's say we have timestamps equivalent to 3.1,3.2 and 3.3 micros. When we read all are rounded to 3 micros. Let's say the last one we read was 3.2. Next time we will read 3.1 again because `3.1 > ROUND(3.2)`

Reply:
![image](https://github.com/user-attachments/assets/c6ff8231-67fa-4f47-a0fc-2f87a278a159)

Will definitely take up this enhancement in follow up PR

Reply:
Can you please add a TODO with JIRA for all the enhancements in the code?

Reply:
Can you test on a big table? I believe we still plan to retain 7 days of data (correct me if I am wrong) and this can be quite a lot of records. And poll would probably most times return 0 records

Reply:
Is it okay if I make the changes related to table indexing & the related tests as part of a follow up PR? There are a few metrics related to initial topic table creations as well which I need to check.

With the current PR, most of the messaging service related changes are implemented and then I plan to use these changes in existing probers to capture the exact stats.

Please let me know your thoughts.

Reply:
PK should be on ts, sequence_id, payload_sequence_id. When I did the experiments, we were hitting the index as long as ts, sequence_id were being provided in the where clause in the same order as PK had been defined. 
AFAIR, Arjan also did the test and saw no issue in it. 
+1 to test it on a large populated table again to determine the latency one more time.

Reply:
SG to test later